server/bytestream/bytestream.go

```diff
@@ -12,13 +13,101 @@ import (
+	"github.com/buildbuddy-io/buildbuddy/server/util/ziputil"
 	"google.golang.org/protobuf/proto"
 
 	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	zipb "github.com/buildbuddy-io/buildbuddy/proto/zip"
 	bspb "google.golang.org/genproto/googleapis/bytestream"
 )
 
-func StreamBytestreamFile(ctx context.Context, env environment.Env, url *url.URL, callback func([]byte)) error {
+func FetchBytestreamZipManifest(ctx context.Context, env environment.Env, url *url.URL) (*zipb.ZipManifest, error) {
+	r, err := digest.ParseDownloadResourceName(strings.TrimPrefix(url.RequestURI(), "/"))
+	if err != nil {
+		return nil, err
+	}
+	// Let's just read 64K and see if we can find the central directory in there.
+	// We probably don't want to be in the business of rendering 3000-file zips'
+	// contents anyway.
+	offset := r.GetDigest().GetSizeBytes() - 65536
+	if offset < 0 {
+		offset = 0
+	}
+
+	var buf bytes.Buffer
+	err = StreamBytestreamFileChunk(ctx, env, url, offset, r.GetDigest().GetSizeBytes()-offset, &buf)
+	if err != nil {
+		return nil, err
+	}
+
+	// Dump the full contents out into a buffer (should be 64K or less).
+	footer := buf.Bytes()
+
+	// Find and parse the End of Central Directory header or fail.
+	eocd, err := ziputil.ReadDirectoryEnd(footer, r.GetDigest().GetSizeBytes())
+	if err != nil {
+		return nil, err
+	}
+
+	cdStart := eocd.DirectoryOffset - offset
+	cdEnd := cdStart + eocd.DirectorySize
+
+	entries, err := ziputil.ReadDirectoryHeader(footer[cdStart:cdEnd], eocd)
+	if err != nil {
+		return nil, err
+	}
+
+	return &zipb.ZipManifest{Entry: entries}, nil
+}
+
+func validateLocalFileHeader(ctx context.Context, env environment.Env, url *url.URL, entry *zipb.ZipManifestEntry) (int, error) {
+	var buf bytes.Buffer
+	err := StreamBytestreamFileChunk(ctx, env, url, entry.GetHeaderOffset(), ziputil.FileHeaderLen, &buf)
+	if err != nil {
+		return -1, err
+	}
+	return ziputil.ValidateLocalFileHeader(buf.Bytes(), entry)
+}
+
+func StreamSingleFileFromBytestreamZip(ctx context.Context, env environment.Env, url *url.URL, entry *zipb.ZipManifestEntry, out io.Writer) error {
+	dynamicHeaderBytes, err := validateLocalFileHeader(ctx, env, url, entry)
+	if err != nil {
+		return err
+	}
+
+	// Stream the dynamic portion of the header and the compressed file as one read.
+	reader, writer := io.Pipe()
+	defer reader.Close()
+	go func() {
+		err := StreamBytestreamFileChunk(ctx, env, url, entry.GetHeaderOffset()+ziputil.FileHeaderLen, entry.GetCompressedSize()+int64(dynamicHeaderBytes), writer)
+		// StreamBytestreamFileChunk shouldn't return EOF, but let's just be safe.
+		if err != nil && err != io.EOF {
+			writer.CloseWithError(err)
+		}
+	}()
+
+	// Validate that the dynamic portion of the file header makes sense.
+	extras := make([]byte, dynamicHeaderBytes)
+	if _, err := io.ReadFull(reader, extras); err != nil {
+		return err
+	}
+	if err := ziputil.ValidateLocalFileNameAndExtras(extras, entry); err != nil {
+		return err
+	}
```

Comment:
if either of these return an error do you want to call reader.CloseWithError() so the goroutine you started dies?

Reply:
I missed the defer reader.Close above -- I think that should be sufficient

Reply:
i have `defer reader.Close()` up where the pipe is opened--once the reader is closed, the writer will get back an `ErrPipeClosed` for subsequent writes, so this should all be good to go already.